### ISSUE TITLE
The Agent needs to be restarted to retrieve data from MQTT

### ISSUE THREAD

**Issue by filgiuff** [telefonicaid/iotagent-json#455]
I've tried the interaction between the MQTT broker and the IoTAgent-JSON.
In a fresh deploy, I created a service and provisioned a device on the Agent:

```
{
	"device_id": "test3",
	"service": "test",
	"service_path": "/test_environment",
	"entity_name": "test3",
	"entity_type": "Device",
	"transport": "MQTT",
	"attributes": [{
		"object_id": "status",
		"name": "status",
		"type": "text"
	}],
	"lazy": [],
	"commands": [],
	"static_attributes": [{
		"name": "location",
		"type": "geo:point",
		"value": "2.424432043119737,6.725294509827826"
	}],
	"protocol": "JSON"
}

```

I sent a measure using the MQTT broker on the topic **/apikeyvalue/test3/attrs** and payload 
```
{"status":"free"}
```
and the entity has been correctly updated on Orion
```
   {
        "id": "test3",
        "type": "Device",
        "TimeInstant": {
            "type": "ISO8601",
            "value": "2019-11-21T16:40:31.00Z",
            "metadata": {}
        },
        "location": {
            "type": "geo:point",
            "value": "2.424432043119737,6.725294509827826",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-11-21T16:40:31.834Z"
                }
            }
        },
        "status": {
            "type": "text",
            "value": "free",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-11-21T16:40:31.834Z"
                }
            }
        }
    }
```
after some time (I still haven't figured out how much, maybe after 20 minutes), if I repeat the same operation (sending a payload on the same MQTT topic) it doesn't work and and no message appears in the log.

**This behaviour happens if there is no device that sends measures for a given time interval. If there is at least one device that continues to send measures in small intervals (for example 9 seconds), the communication works for all devices.**
The devices that are not MQTT continue to work properly.

**Restarting the IoTAgent everything returns to work properly (until the above conditions are repeated).**

I've tried with:

-  Orion 1.13.0 and IoTAgent-JSON 1.6.0
-  Orion 2.3.0 and IoTAgent-JSON 1.12.0

The same behaviour happens with the IoTAgent-UL agent.

**Comment by fgalan** [comment]
MQTT transport has been quite improved in last version released last week (1.13.0). I'd suggest to give a try to the set up with:

* Orion 2.3.0 and IoTAgent-JSON 1.13.0

and tell us how it goes, please.

Thanks!

**Comment by filgiuff** [comment]
@fgalan 

I've just tried with these versions (Orion 2.3.0 and IoTAgent-JSON 1.13.0) and in this case it doesn't work immediately.
```
{
    "devices": [{
            "device_id": "test8",
            "service": "test",
            "service_path": "/test_environment",
            "entity_name": "test8",
            "entity_type": "Device",
            "transport": "MQTT",
            "attributes": [
                {
                    "object_id": "status",
                    "name": "status",
                    "type": "text"
                }
            ],
            "lazy": [],
            "commands": [],
            "static_attributes": [
                {
                    "name": "location",
                    "type": "geo:point",
                    "value": "-4.424432043119737,36.725294509827826"
                }
            ],
            "protocol": "JSON"
        }]
}
```


I've tried to use **/apikeyvalue/test8/attrs** (with "body" {"status":"free"}) and also **/apikeyvalue/test8/attrs/status** (with "body" free) but the measure is always empty.

```
{
        "id": "test8",
        "type": "Device",
        "TimeInstant": {
            "type": "ISO8601",
            "value": " ",
            "metadata": {}
        },
        "location": {
            "type": "geo:point",
            "value": "-4.424432043119737,36.725294509827826",
            "metadata": {}
        },
        "status": {
            "type": "text",
            "value": " ",
            "metadata": {}
        }
    }
```

Did I make any mistakes?
I'm using eclipse-mosquitto:1.4.12, everything is deployed through Docker.

**Comment by fgalan** [comment]
>  I'm using eclipse-mosquitto:1.4.12, everything is deployed through Docker.

Looking to https://github.com/telefonicaid/iotagent-json/releases/tag/1.13.0 I read:

>  Needs MQTT v5 broker like mosquitto 1.6+

So maybe it should be upgraded. To be precise, the one used by Travis CI is eclipse-mosquitto:1.6.7 (see https://github.com/telefonicaid/iotagent-json/blob/1.13.0/.travis.yml#L32)

**Comment by filgiuff** [comment]
I've just tried with eclipse-mosquitto:1.6.7.
Immediately after the deploy the measure has been updated using both /apikeyvalue/test8/attrs and /json/apikeyvalue/test8/attrs  with "body" {"status":"free"}
But trying again after more or less 25 minutes (no measures were sent via MQTT in this period) the measure has not been updated. Restarting the agent, it returns to work properly (until the above conditions are repeated).
 
In other words the general issue of this topic is still present.

**Comment by AlvaroVega** [comment]
Thanks, we will check it!

**Comment by AlvaroVega** [comment]
I've just tried with iotagent-ul sending measures after 25 minutes and worked.

**Comment by fgalan** [comment]
> But trying again after more or less 25 minutes (no measures were sent via MQTT in this period) the measure has not been updated. 

In the case you periodically send a measure with a shorter interval (i.e. every 5 minutes), does the behavior change? Is it still works after the first 25 minutes?

**Comment by filgiuff** [comment]
> 
> 
> I've just tried with iotagent-ul sending measures after 25 minutes and worked.

I've just tried using iotagent-ul:1.12, eclipse-mosquitto:1.6.7 and orion 2.3.0 and after 2 hours it doesn't work.

**Comment by filgiuff** [comment]
> 
> 
> > But trying again after more or less 25 minutes (no measures were sent via MQTT in this period) the measure has not been updated.
> 
> In the case you periodically send a measure with a shorter interval (i.e. every 5 minutes), does the behavior change? Is it still works after the first 25 minutes?

This is the workaround that we are using to avoid the issue: we have a job that works as a fake device that sends a measure to the MQTT broker (in a topic of a device provisioned on the agent) every 10 seconds, updating also Orion. While this job is active everything works fine, but it remains anyway a workaround.

**Comment by AlvaroVega** [comment]
Could you provide us iotagent traces (debug level) ?

**Comment by filgiuff** [comment]
Here they are
[iotagent-json-log.zip](https://github.com/telefonicaid/iotagent-json/files/3897319/iotagent-json-log.zip) Log of the agent 
[docker-compose.zip](https://github.com/telefonicaid/iotagent-json/files/3897322/docker-compose.zip) docker compose with configuration
[mosquittos.zip](https://github.com/telefonicaid/iotagent-json/files/3897320/mosquittos.zip) mosquitto (s) configuration files

I only hid my public IP using ###.###.###.###

**Comment by AlvaroVega** [comment]
I've se you use:
      - "IOTA_MQTT_QOS=0"
      - "IOTA_MQTT_RETAIN=false"
and I'm used to use:
      - "IOTA_MQTT_QOS=2"
      - "IOTA_MQTT_RETAIN=true"

**Comment by fgalan** [comment]
> 
> 
> I've see you use:
> - "IOTA_MQTT_QOS=0"
> - "IOTA_MQTT_RETAIN=false"
> and I'm used to use:
> - "IOTA_MQTT_QOS=2"
> - "IOTA_MQTT_RETAIN=true"
> I'm trying to do a long test with your values..

Maybe @filgiuff you could do the cross-test (I mean, to use the same values used by @AlvaroVega and see how it goes)

**Comment by filgiuff** [comment]
@fgalan I took your advice and did a test using the same values used by @AlvaroVega, but the problem remains.

**Comment by filgiuff** [comment]
Hi all,
I have a good news, finally I found the problem!
I set this configuration:

- IOTA_MQTT_QOS=0
- IOTA_MQTT_RETAIN=false
- **IOTA_MQTT_KEEPALIVE=60**

I tried it after 2 days and it worked fine!
This IOTA_MQTT_KEEPALIVE solved the issue!

**Comment by fgalan** [comment]
@filgiuff happy to know you finally solved! :)

Just to know, before setting IOTA_MQTT_KEEPALIVE=60 which value do you have for that env var? Or maybe the env var was not defined at all?

**Comment by filgiuff** [comment]
Using  version 1.12.0 it was not defined at all.
Using 1.6.0 it's not possible to define this variable, so I downloaded the code, changed this value directly on the code and now it works fine.

**Comment by fgalan** [comment]
Looking to how IOTA_MQTT_KEEPALIVE is used internally, it is used to set the mqtt.keepalive configuration parameter. If the variable is not set the default mqtt.keepalive in config.js is used (default is 0). If mqtt.keepalive is not used in config.js, 0 is used also.

https://github.com/telefonicaid/iotagent-json/blob/52b971a79e6e304bdbd4b3a26898a60fd1deed20/lib/bindings/MQTTBinding.js#L183-L185

This is used to do the MQTT connection:

https://github.com/telefonicaid/iotagent-json/blob/52b971a79e6e304bdbd4b3a26898a60fd1deed20/lib/bindings/MQTTBinding.js#L207-L210

I understand that 0 means "never send a keepalive" so maybe the underlaying networking in the implementation can cause disconnect in this case (that could explain why the situation is different in your evironment and in @AlvaroVega enviroments, although both of you are using the same configuration).

Maybe a recommendation of using greater than 0 value for this parameter in the case of having disconnect problems should be included in the documentation. What do you think?

**Comment by filgiuff** [comment]
> Maybe a recommendation of using greater than 0 value for this parameter in the case of having disconnect problems should be included in the documentation. What do you think?

I totally agree.

As final user, based on my experience with using this agent, I can suggest you to set a keepAlive default value greater than 0.

My main problem was that I was using the version 1.6.0 where the keepAlive value is not configurable, using last releases this problem is less serious.

Thanks.

**Comment by fgalan** [comment]
PR https://github.com/telefonicaid/iotagent-json/pull/460 adjust documentation.

Please @filgiuff have a look and if you find it ok, please close the issue.
